// dockerfile-language-server-nodejs/lib/parser/instructions/copy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Remy Suen. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
const vscode_languageserver_1 = require("vscode-languageserver");
const jsonInstruction_1 = require("./jsonInstruction");
class Copy extends jsonInstruction_1.JSONInstruction {
    constructor(document, range, escapeChar, instruction, instructionRange) {
        super(document, range, escapeChar, instruction, instructionRange);
    }
    getArgumentsContent() {
        const args = this.getArguments();
        const flags = this.getFlags();
        if (args.length === 0 || flags.length === 0) {
            return super.getArgumentsContent();
        }
        return this.getRangeContent(vscode_languageserver_1.Range.create(args[0].getRange().start, args[args.length - 1].getRange().end));
    }
    stopSearchingForFlags(argument) {
        return argument.indexOf("--") === -1;
    }
    getFromFlag() {
        let flags = super.getFlags();
        return flags.length === 1 && flags[0].getName() === "from" ? flags[0] : null;
    }
    getFromValue() {
        return this.getRangeContent(this.getFromValueRange());
    }
    getFromValueRange() {
        const flag = this.getFromFlag();
        return flag ? flag.getValueRange() : null;
    }
}
exports.Copy = Copy;
