// dockerfile-language-server-nodejs/lib/parser/instructions/jsonInstruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Remy Suen. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
const vscode_languageserver_1 = require("vscode-languageserver");
const argument_1 = require("../argument");
const modifiableInstruction_1 = require("./modifiableInstruction");
class JSONInstruction extends modifiableInstruction_1.ModifiableInstruction {
    constructor(document, range, escapeChar, instruction, instructionRange) {
        super(document, range, escapeChar, instruction, instructionRange);
        this.jsonStrings = [];
        const argsContent = this.getArgumentsContent();
        if (argsContent === null) {
            return;
        }
        const args = this.getArguments();
        if (args.length === 1 && args[0].getValue() === "[]") {
            let argRange = args[0].getRange();
            this.openingBracket = new argument_1.Argument("[", vscode_languageserver_1.Range.create(argRange.start.line, argRange.start.character, argRange.start.line, argRange.start.character + 1));
            this.closingBracket = new argument_1.Argument("]", vscode_languageserver_1.Range.create(argRange.start.line, argRange.start.character + 1, argRange.end.line, argRange.end.character));
            return;
        }
        else if (args.length === 2 && args[0].getValue() === '[' && args[1].getValue() === ']') {
            this.openingBracket = args[0];
            this.closingBracket = args[1];
            return;
        }
        const argsOffset = document.offsetAt(this.getArgumentsRange().start);
        let fullStart = -1;
        let last = "";
        let quoted = false;
        argsCheck: for (let i = 0; i < argsContent.length; i++) {
            switch (argsContent.charAt(i)) {
                case '[':
                    if (last === "") {
                        this.openingBracket = new argument_1.Argument("[", vscode_languageserver_1.Range.create(document.positionAt(argsOffset + i), document.positionAt(argsOffset + i + 1)));
                        last = '[';
                        fullStart = i + 1;
                    }
                    else if (!quoted) {
                        break argsCheck;
                    }
                    break;
                case '"':
                    if (last === '[' || last === ',') {
                        quoted = true;
                        last = '"';
                        continue;
                    }
                    else if (last === '"') {
                        if (quoted) {
                            // quoted string done
                            quoted = false;
                        }
                        else {
                            // should be a , or a ]
                            break argsCheck;
                        }
                    }
                    else {
                        break argsCheck;
                    }
                    break;
                case ',':
                    if (!quoted) {
                        this.jsonStrings.push(new argument_1.Argument(argsContent.substring(fullStart, i), vscode_languageserver_1.Range.create(document.positionAt(argsOffset + fullStart), document.positionAt(argsOffset + i))));
                        fullStart = i + 1;
                        if (last === '"') {
                            last = ',';
                        }
                        else {
                            break argsCheck;
                        }
                    }
                    break;
                case ']':
                    if (!quoted && last !== "") {
                        this.jsonStrings.push(new argument_1.Argument(argsContent.substring(fullStart, i), vscode_languageserver_1.Range.create(document.positionAt(argsOffset + fullStart), document.positionAt(argsOffset + i))));
                        this.closingBracket = new argument_1.Argument("]", vscode_languageserver_1.Range.create(document.positionAt(argsOffset + i), document.positionAt(argsOffset + i + 1)));
                        break argsCheck;
                    }
                    break;
                case ' ':
                case '\t':
                    break;
                case '\\':
                    if (quoted) {
                        switch (argsContent.charAt(i + 1)) {
                            case '"':
                            case '\\':
                                i++;
                                continue;
                            case ' ':
                            case '\t':
                                for (let j = i + 2; j < argsContent.length; j++) {
                                    switch (argsContent.charAt(j)) {
                                        case '\r':
                                            if (argsContent.charAt(j + 1) === '\n') {
                                                j++;
                                            }
                                        case '\n':
                                            i = j;
                                            continue argsCheck;
                                        case ' ':
                                        case '\t':
                                            break;
                                        default:
                                            break argsCheck;
                                    }
                                }
                                break;
                            default:
                                i++;
                                continue;
                        }
                    }
                    else {
                        for (let j = i + 1; j < argsContent.length; j++) {
                            switch (argsContent.charAt(j)) {
                                case '\r':
                                    if (argsContent.charAt(j + 1) === '\n') {
                                        j++;
                                    }
                                case '\n':
                                    i = j;
                                    continue argsCheck;
                                case ' ':
                                case '\t':
                                    break;
                                default:
                                    break argsCheck;
                            }
                        }
                    }
                    break;
                default:
                    if (!quoted) {
                        break argsCheck;
                    }
                    break;
            }
        }
    }
    stopSearchingForFlags(_value) {
        return true;
    }
    getOpeningBracket() {
        return this.openingBracket;
    }
    getJSONStrings() {
        return this.jsonStrings;
    }
    getClosingBracket() {
        return this.closingBracket;
    }
}
exports.JSONInstruction = JSONInstruction;
